File: KEM/bindings/swift/Sources/KazKem/KazKem.swift
import Foundation
import CKazKem
import Security

/// KAZ-KEM Post-Quantum Key Encapsulation Mechanism.
///
/// Thread-safe wrapper for the native KAZ-KEM library providing
/// post-quantum secure key encapsulation.
///
/// ## Usage
/// ```swift
/// // Initialize with security level
/// try KazKem.initialize(level: .level128)
///
/// // Generate key pair
/// let keyPair = try KazKem.generateKeyPair()
///
/// // Encapsulate shared secret
/// let result = try KazKem.encapsulate(publicKey: keyPair.getPublicKey())
///
/// // Decapsulate on recipient side
/// let sharedSecret = try KazKem.decapsulate(ciphertext: result.ciphertext, privateKey: keyPair.privateKey)
/// ```
public final class KazKem: @unchecked Sendable {

    // MARK: - Private Properties

    private static let lock = NSLock()
    private static var _current: KazKem?

    private let _securityLevel: SecurityLevel
    private let _publicKeySize: Int
    private let _privateKeySize: Int
    private let _ciphertextSize: Int
    private let _sharedSecretSize: Int

    // MARK: - Public Properties

    /// Current security level
    public var securityLevel: SecurityLevel { _securityLevel }

    /// Public key size in bytes
    public var publicKeySize: Int { _publicKeySize }

    /// Private key size in bytes
    public var privateKeySize: Int { _privateKeySize }

    /// Ciphertext size in bytes
    public var ciphertextSize: Int { _ciphertextSize }

    /// Shared secret size in bytes
    public var sharedSecretSize: Int { _sharedSecretSize }

    /// Library version string
    public static var version: String {
        guard let ptr = kaz_kem_version() else { return "unknown" }
        return String(cString: ptr)
    }

    /// Check if KAZ-KEM is initialized
    public static var isInitialized: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _current != nil && kaz_kem_is_initialized() != 0
    }

    /// Get the current initialized instance
    public static var current: KazKem {
        get throws {
            lock.lock()
            defer { lock.unlock() }

            guard let instance = _current else {
                throw KazKemError.notInitialized
            }
            return instance
        }
    }

    // MARK: - Initialization

    private init(level: SecurityLevel) {
        self._securityLevel = level
        self._publicKeySize = Int(kaz_kem_publickey_bytes())
        self._privateKeySize = Int(kaz_kem_privatekey_bytes())
        self._ciphertextSize = Int(kaz_kem_ciphertext_bytes())
        self._sharedSecretSize = Int(kaz_kem_shared_secret_bytes())
    }

    /// Initialize KAZ-KEM with the specified security level.
    ///
    /// - Parameter level: Security level (128, 192, or 256 bits)
    /// - Returns: Initialized KAZ-KEM instance
    /// - Throws: `KazKemError` if initialization fails
    @discardableResult
    public static func initialize(level: SecurityLevel = .level128) throws -> KazKem {
        lock.lock()
        defer { lock.unlock() }

        // If already initialized with same level, return existing
        if let current = _current, current._securityLevel == level {
            return current
        }

        // Cleanup previous instance
        if _current != nil {
            kaz_kem_cleanup()
            _current = nil
        }

        let result = kaz_kem_init(Int32(level.rawValue))
        if result != 0 {
            throw KazKemError.from(code: result, operation: "initialize")
        }

        let instance = KazKem(level: level)
        _current = instance
        return instance
    }

    /// Cleanup and release resources.
    public static func cleanup() {
        lock.lock()
        defer { lock.unlock() }

        if _current != nil {
            kaz_kem_cleanup()
            _current = nil
        }
    }

    // MARK: - Key Generation

    /// Generate a new key pair.
    ///
    /// - Returns: A new key pair containing public and private keys
    /// - Throws: `KazKemError` if key generation fails
    public func generateKeyPair() throws -> KazKemKeyPair {
        try ensureInitialized()

        var publicKey = [UInt8](repeating: 0, count: publicKeySize)
        var privateKey = [UInt8](repeating: 0, count: privateKeySize)

        let result = kaz_kem_keypair(&publicKey, &privateKey)
        if result != 0 {
            // Clear partial data
            secureZero(&privateKey)
            throw KazKemError.from(code: result, operation: "generateKeyPair")
        }

        return KazKemKeyPair(
            publicKey: Data(publicKey),
            privateKey: Data(privateKey),
            securityLevel: securityLevel
        )
    }

    /// Static convenience method to generate a key pair.
    public static func generateKeyPair() throws -> KazKemKeyPair {
        return try current.generateKeyPair()
    }

    // MARK: - Encapsulation

    /// Encapsulate a shared secret using the recipient's public key.
    ///
    /// - Parameter publicKey: Recipient's public key
    /// - Returns: Encapsulation result containing ciphertext and shared secret
    /// - Throws: `KazKemError` if encapsulation fails
    public func encapsulate(publicKey: KazKemPublicKey) throws -> KazKemEncapsulationResult {
        return try encapsulate(publicKey: publicKey.data)
    }

    /// Encapsulate a shared secret using the recipient's public key bytes.
    ///
    /// - Parameter publicKey: Recipient's public key as Data
    /// - Returns: Encapsulation result containing ciphertext and shared secret
    /// - Throws: `KazKemError` if encapsulation fails
    public func encapsulate(publicKey: Data) throws -> KazKemEncapsulationResult {
        try ensureInitialized()

        guard publicKey.count == publicKeySize else {
            throw KazKemError.invalidParameter(
                "Public key must be \(publicKeySize) bytes, got \(publicKey.count)"
            )
        }

        // Generate random shared secret
        var sharedSecret = [UInt8](repeating: 0, count: sharedSecretSize)
        let status = SecRandomCopyBytes(kSecRandomDefault, sharedSecretSize, &sharedSecret)
        guard status == errSecSuccess else {
            throw KazKemError.randomGenerationFailed
        }

        // Ensure message < N by clearing high bits based on security level
        sharedSecret[0] &= securityLevel.randomMask

        var ciphertext = [UInt8](repeating: 0, count: ciphertextSize)
        var ctLen: UInt64 = 0

        let result = publicKey.withUnsafeBytes { pkPtr -> Int32 in
            return kaz_kem_encapsulate(
                &ciphertext,
                &ctLen,
                sharedSecret,
                UInt64(sharedSecret.count),
                pkPtr.baseAddress?.assumingMemoryBound(to: UInt8.self)
            )
        }

        if result != 0 {
            secureZero(&sharedSecret)
            throw KazKemError.from(code: result, operation: "encapsulate")
        }

        return KazKemEncapsulationResult(
            ciphertext: Data(ciphertext.prefix(Int(ctLen))),
            sharedSecret: Data(sharedSecret)
        )
    }

    /// Static convenience method to encapsulate.
    public static func encapsulate(publicKey: KazKemPublicKey) throws -> KazKemEncapsulationResult {
        return try current.encapsulate(publicKey: publicKey)
    }

    /// Static convenience method to encapsulate.
    public static func encapsulate(publicKey: Data) throws -> KazKemEncapsulationResult {
        return try current.encapsulate(publicKey: publicKey)
    }

    // MARK: - Decapsulation

    /// Decapsulate a shared secret using the private key.
    ///
    /// - Parameters:
    ///   - ciphertext: Ciphertext from encapsulation
    ///   - keyPair: Key pair containing the private key
    /// - Returns: The shared secret
    /// - Throws: `KazKemError` if decapsulation fails
    public func decapsulate(ciphertext: Data, keyPair: KazKemKeyPair) throws -> Data {
        return try decapsulate(ciphertext: ciphertext, privateKey: keyPair.privateKey)
    }

    /// Decapsulate a shared secret using the private key bytes.
    ///
    /// - Parameters:
    ///   - ciphertext: Ciphertext from encapsulation
    ///   - privateKey: Private key as Data
    /// - Returns: The shared secret
    /// - Throws: `KazKemError` if decapsulation fails
    public func decapsulate(ciphertext: Data, privateKey: Data) throws -> Data {
        try ensureInitialized()

        guard privateKey.count == privateKeySize else {
            throw KazKemError.invalidParameter(
                "Private key must be \(privateKeySize) bytes, got \(privateKey.count)"
            )
        }

        guard ciphertext.count > 0 && ciphertext.count <= ciphertextSize else {
            throw KazKemError.invalidParameter(
                "Ciphertext must be 1-\(ciphertextSize) bytes, got \(ciphertext.count)"
            )
        }

        var sharedSecret = [UInt8](repeating: 0, count: sharedSecretSize)
        var ssLen: UInt64 = 0

        let result = ciphertext.withUnsafeBytes { ctPtr -> Int32 in
            privateKey.withUnsafeBytes { skPtr -> Int32 in
                return kaz_kem_decapsulate(
                    &sharedSecret,
                    &ssLen,
                    ctPtr.baseAddress?.assumingMemoryBound(to: UInt8.self),
                    UInt64(ciphertext.count),
                    skPtr.baseAddress?.assumingMemoryBound(to: UInt8.self)
                )
            }
        }

        if result != 0 {
            secureZero(&sharedSecret)
            throw KazKemError.from(code: result, operation: "decapsulate")
        }

        return Data(sharedSecret.prefix(Int(ssLen)))
    }

    /// Static convenience method to decapsulate.
    public static func decapsulate(ciphertext: Data, keyPair: KazKemKeyPair) throws -> Data {
        return try current.decapsulate(ciphertext: ciphertext, keyPair: keyPair)
    }

    /// Static convenience method to decapsulate.
    public static func decapsulate(ciphertext: Data, privateKey: Data) throws -> Data {
        return try current.decapsulate(ciphertext: ciphertext, privateKey: privateKey)
    }

    // MARK: - Private Helpers

    private func ensureInitialized() throws {
        guard kaz_kem_is_initialized() != 0 else {
            throw KazKemError.notInitialized
        }
    }
}

// MARK: - Extensions

extension Data {
    /// Convert to hexadecimal string
    public var hexString: String {
        return map { String(format: "%02x", $0) }.joined()
    }

    /// Initialize from hexadecimal string
    public init?(hexString: String) {
        let hex = hexString.lowercased()
        guard hex.count % 2 == 0 else { return nil }

        var data = Data()
        var index = hex.startIndex
        while index < hex.endIndex {
            let nextIndex = hex.index(index, offsetBy: 2)
            guard let byte = UInt8(hex[index..<nextIndex], radix: 16) else { return nil }
            data.append(byte)
            index = nextIndex
        }
        self = data
    }
}

File: KEM/bindings/android/kazkem/src/main/kotlin/com/antrapol/kaz/kem/KazKem.kt
package com.antrapol.kaz.kem

import androidx.annotation.GuardedBy
import java.security.SecureRandom
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * KAZ-KEM Post-Quantum Key Encapsulation Mechanism.
 *
 * Thread-safe wrapper for the native KAZ-KEM library providing
 * post-quantum secure key encapsulation.
 *
 * ## Usage
 * ```kotlin
 * // Initialize with security level
 * val kem = KazKem.initialize(SecurityLevel.LEVEL_128)
 *
 * // Generate key pair
 * val keyPair = kem.generateKeyPair()
 *
 * // Encapsulate shared secret (sender side)
 * val result = kem.encapsulate(keyPair.getPublicKey())
 *
 * // Decapsulate on recipient side
 * val sharedSecret = kem.decapsulate(result.ciphertext, keyPair.privateKey)
 * ```
 *
 * ## Thread Safety
 * This class is thread-safe. Multiple threads can safely generate keys and
 * perform encapsulation/decapsulation operations concurrently.
 */
class KazKem private constructor(
    private val _securityLevel: SecurityLevel
) {
    /**
     * Current security level.
     */
    val securityLevel: SecurityLevel
        get() = _securityLevel

    /**
     * Public key size in bytes for the current security level.
     */
    val publicKeySize: Int
        get() = KazKemNative.nativePublicKeyBytes()

    /**
     * Private key size in bytes for the current security level.
     */
    val privateKeySize: Int
        get() = KazKemNative.nativePrivateKeyBytes()

    /**
     * Ciphertext size in bytes for the current security level.
     */
    val ciphertextSize: Int
        get() = KazKemNative.nativeCiphertextBytes()

    /**
     * Shared secret size in bytes for the current security level.
     */
    val sharedSecretSize: Int
        get() = KazKemNative.nativeSharedSecretBytes()

    /**
     * Generate a new key pair.
     *
     * @return A new [KazKemKeyPair] containing public and private keys
     * @throws KazKemException if key generation fails
     * @throws NotInitializedException if KAZ-KEM is not initialized
     */
    @Throws(KazKemException::class, NotInitializedException::class)
    fun generateKeyPair(): KazKemKeyPair {
        ensureInitialized()

        val result = KazKemNative.nativeKeyPair()
        return KazKemKeyPair(result[0], result[1], securityLevel)
    }

    /**
     * Encapsulate a shared secret using the recipient's public key.
     *
     * @param publicKey Recipient's public key
     * @return [KazKemEncapsulationResult] containing ciphertext and shared secret
     * @throws KazKemException if encapsulation fails
     * @throws InvalidParameterException if public key size is invalid
     * @throws NotInitializedException if KAZ-KEM is not initialized
     */
    @Throws(KazKemException::class, InvalidParameterException::class, NotInitializedException::class)
    fun encapsulate(publicKey: KazKemPublicKey): KazKemEncapsulationResult {
        return encapsulate(publicKey.data)
    }

    /**
     * Encapsulate a shared secret using the recipient's public key bytes.
     *
     * @param publicKey Recipient's public key as ByteArray
     * @return [KazKemEncapsulationResult] containing ciphertext and shared secret
     * @throws KazKemException if encapsulation fails
     * @throws InvalidParameterException if public key size is invalid
     * @throws NotInitializedException if KAZ-KEM is not initialized
     */
    @Throws(KazKemException::class, InvalidParameterException::class, NotInitializedException::class)
    fun encapsulate(publicKey: ByteArray): KazKemEncapsulationResult {
        ensureInitialized()

        if (publicKey.size != publicKeySize) {
            throw InvalidParameterException(
                "Public key must be $publicKeySize bytes, got ${publicKey.size}"
            )
        }

        // Generate random shared secret
        val sharedSecret = ByteArray(sharedSecretSize)
        SecureRandom().nextBytes(sharedSecret)

        // Ensure message < N by clearing high bits based on security level
        sharedSecret[0] = (sharedSecret[0].toInt() and securityLevel.randomMask.toInt()).toByte()

        val result = KazKemNative.nativeEncapsulate(sharedSecret, publicKey)
        return KazKemEncapsulationResult(result[0], sharedSecret)
    }

    /**
     * Decapsulate a shared secret using the private key.
     *
     * @param ciphertext Ciphertext from encapsulation
     * @param keyPair Key pair containing the private key
     * @return The shared secret as ByteArray
     * @throws KazKemException if decapsulation fails
     * @throws NotInitializedException if KAZ-KEM is not initialized
     */
    @Throws(KazKemException::class, NotInitializedException::class)
    fun decapsulate(ciphertext: ByteArray, keyPair: KazKemKeyPair): ByteArray {
        return decapsulate(ciphertext, keyPair.privateKey)
    }

    /**
     * Decapsulate a shared secret using the private key bytes.
     *
     * @param ciphertext Ciphertext from encapsulation
     * @param privateKey Private key as ByteArray
     * @return The shared secret as ByteArray
     * @throws KazKemException if decapsulation fails
     * @throws InvalidParameterException if key/ciphertext size is invalid
     * @throws NotInitializedException if KAZ-KEM is not initialized
     */
    @Throws(KazKemException::class, InvalidParameterException::class, NotInitializedException::class)
    fun decapsulate(ciphertext: ByteArray, privateKey: ByteArray): ByteArray {
        ensureInitialized()

        if (privateKey.size != privateKeySize) {
            throw InvalidParameterException(
                "Private key must be $privateKeySize bytes, got ${privateKey.size}"
            )
        }

        if (ciphertext.isEmpty() || ciphertext.size > ciphertextSize) {
            throw InvalidParameterException(
                "Ciphertext must be 1-$ciphertextSize bytes, got ${ciphertext.size}"
            )
        }

        return KazKemNative.nativeDecapsulate(ciphertext, privateKey)
    }

    private fun ensureInitialized() {
        if (!KazKemNative.nativeIsInitialized()) {
            throw NotInitializedException()
        }
    }

    companion object {
        private val lock = ReentrantLock()

        @GuardedBy("lock")
        private var _current: KazKem? = null

        /**
         * Library version string.
         */
        @JvmStatic
        val version: String
            get() = KazKemNative.nativeVersion()

        /**
         * Check if KAZ-KEM is initialized.
         */
        @JvmStatic
        val isInitialized: Boolean
            get() = lock.withLock { _current != null && KazKemNative.nativeIsInitialized() }

        /**
         * Get the current initialized instance.
         *
         * @throws NotInitializedException if not initialized
         */
        @JvmStatic
        @get:Throws(NotInitializedException::class)
        val current: KazKem
            get() = lock.withLock {
                _current ?: throw NotInitializedException()
            }

        /**
         * Initialize KAZ-KEM with the specified security level.
         *
         * @param level Security level (default: [SecurityLevel.LEVEL_128])
         * @return Initialized [KazKem] instance
         * @throws KazKemException if initialization fails
         */
        @JvmStatic
        @JvmOverloads
        @Throws(KazKemException::class)
        fun initialize(level: SecurityLevel = SecurityLevel.LEVEL_128): KazKem {
            return lock.withLock {
                // If already initialized with same level, return existing
                _current?.let { current ->
                    if (current._securityLevel == level) {
                        return@withLock current
                    }
                }

                // Cleanup previous instance
                if (_current != null) {
                    KazKemNative.nativeCleanup()
                    _current = null
                }

                // Initialize with new level
                val result = KazKemNative.nativeInit(level.value)
                if (result != 0) {
                    throw KazKemException.fromErrorCode(result, "initialize")
                }

                val instance = KazKem(level)
                _current = instance
                instance
            }
        }

        /**
         * Cleanup and release resources.
         */
        @JvmStatic
        fun cleanup() {
            lock.withLock {
                if (_current != null) {
                    KazKemNative.nativeCleanup()
                    _current = null
                }
            }
        }
    }
}
